### PR TITLE
[Example] metapath2vec

### PR DESCRIPTION
**Pull request by ziqiaomeng**
## Description
Metapath2vec implementation. Will possibly have sampler optimization in the future. Metapath2vec++ is not ready yet.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by zheng-da** [comment]
Can you write a README for metapath2vec? This includes how people should run the code to reproduce the performance result. I know you made a lot of efforts to reproduce the performance in the paper. It'll be great if you can write what tricks you have tried to reproduce the results.

**Comment by ziqiaomeng** [comment]
No problem. I will write it as soon as possible.

File: examples/pytorch/metapath2vec/reading_data.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+        self.word_count = len(self.word2id)
+        #self.index2type = index2type
+        #self.type2indices = type2indices
+        print("Total embeddings: " + str(len(self.word2id)))
+
+    def initTableDiscards(self):
+        t = 0.0001
+        f = np.array(list(self.word_frequency.values())) / self.token_count
+        self.discards = np.sqrt(t / f) + (t / f)
```
Comment (zheng-da) [inline examples/pytorch/metapath2vec/reading_data.py]: it seems this isn't used. Why do you have this function?

File: examples/pytorch/metapath2vec/reading_data.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+    def initTableNegatives(self):
```
Comment (zheng-da) [inline examples/pytorch/metapath2vec/reading_data.py]: please add comments on this function. What does this function do?

File: examples/pytorch/metapath2vec/reading_data.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+class Metapath2vecDataset(Dataset):
+    def __init__(self, data, window_size):
+        self.data = data
+        self.window_size = window_size
+        self.input_file = open(data.inputFileName, encoding="utf8")
+
+    def __len__(self):
+        return self.data.sentences_count
+
+    def __getitem__(self, idx):
```
Comment (zheng-da) [inline examples/pytorch/metapath2vec/reading_data.py]: please document what is read and what is returned.

File: examples/pytorch/metapath2vec/sampler.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+walk_length = 100
+
+#construct mapping from text, could be changed to DGL later
+def construct_id_dict():
+    id_to_paper = {}
+    id_to_author = {}
+    id_to_conf = {}
+    f_3 = open("/Users/ziqiaom/Desktop/net_dbis/id_author.txt", encoding="ISO-8859-1")
+    f_4 = open("/Users/ziqiaom/Desktop/net_dbis/id_conf.txt", encoding="ISO-8859-1")
+    f_5 = open("/Users/ziqiaom/Desktop/net_dbis/paper.txt", encoding="ISO-8859-1")
```
Comment (zheng-da) [inline examples/pytorch/metapath2vec/sampler.py]: please not use your local path.

File: examples/pytorch/metapath2vec/sampler.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    return id_to_paper, id_to_author, id_to_conf
+
+#construct mapping from text, could be changed to DGL later
+def construct_types_mappings():
+    paper_to_author = {}
+    author_to_paper = {}
+    paper_to_conf = {}
+    conf_to_paper = {}
+    f_1 = open("/Users/ziqiaom/Desktop/net_dbis/paper_author.txt", "r")
+    f_2 = open("/Users/ziqiaom/Desktop/net_dbis/paper_conf.txt", "r")
```
Comment (zheng-da) [inline examples/pytorch/metapath2vec/sampler.py]: the same here.

File: examples/pytorch/metapath2vec/sampler.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        else:
+            conf_to_paper[y[1]] = []
+            conf_to_paper[y[1]].append(y[0])
+    f_1.close()
+    f_2.close()
+    return paper_to_author, author_to_paper, paper_to_conf, conf_to_paper
+
+#"conference - paper - Author - paper - conference" metapath sampling
+def generate_metapath():
+    output_path = open("/Users/ziqiaom/Desktop/net_dbis/output_path.txt", "w")
```
Comment (zheng-da) [inline examples/pytorch/metapath2vec/sampler.py]: the same here.

File: examples/pytorch/metapath2vec/metapath2vec.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+-----
+Run with the following procedures:
+
+1, Run sampler.py on your graph dataset. Note that: the input text file should be list of mappings so you probably need to preprocess your graph dataset. Files with sample format are available in "net_dbis" file. Of course you could also use your own metapath sampler implementation.
+
+2, Run the following command:
+```bash
+python metapath2vec.py --input_file "aminer.txt" --output_file "your_output_file_path"
+```
+aminer.txt file is available here: [https://drive.google.com/drive/my-drive]
```
[the remark this answers is not in the document]
Comment (ziqiaomeng) [reply examples/pytorch/metapath2vec/metapath2vec.md]: I upload aminer.txt file to my google drive. It is a large file with 780 million lines.

**Comment by zheng-da** [comment]
please remove the data files from the PR. Instead, please upload them to S3 and write a data loader that downloads the files from S3.

File: examples/pytorch/metapath2vec/reading_data.py (hunk 2 above)
Comment (ziqiaomeng) [reply examples/pytorch/metapath2vec/reading_data.py]: OK, I will add

**Comment by ziqiaomeng** [comment]
OK, no problem.

File: examples/pytorch/metapath2vec/reading_data.py (hunk 1 above)
Comment (ziqiaomeng) [reply examples/pytorch/metapath2vec/reading_data.py]: This function is called in the very front, preparing a frequency table for sub-sampling.

**Comment by yzh119** [comment]
What's the progress of this PR?

**Comment by yzh119** [comment]
Wait, the name of `metapath2vec.md` should be `README.md`...

**Comment by mrbeann** [comment]
Hi, I read some code in this repo, but it seems the sampler.py is not used. Now, this may be a Deepwalk implementation instead of metapath2vec?

**Comment by ziqiaomeng** [comment]
Thanks for your question. The author’s implementation is also “two-stage”. Firstly sampling on graph dataset and then running learning algorithm on sampled sequences. The data named “aminer.txt” downloaded from aws s3 is actually bunch of sampled sequences.
________________________________
发件人: JC Liu <notifications@github.com>
发送时间: 2019年9月2日 8:19:57
收件人: dmlc/dgl <dgl@noreply.github.com>
抄送: Ziqiao Meng <ziqiao.meng@mail.utoronto.ca>; Author <author@noreply.github.com>
主题: Re: [dmlc/dgl] [Example] metapath2vec (#764)


Hi, I read some code in this repo, but it seems the sampler.py is not used. Now, this may be a Deepwalk implementation instead of metapath2vec?

―

**Comment by ziqiaomeng** [comment]
And also the sampler is not very general. It could only be used on aminer dataset. DGL team might provide sampler optimization in the future. So current version of metapath2vec is not expected to be the terminal version.
________________________________
发件人: JC Liu <notifications@github.com>
发送时间: 2019年9月2日 8:19:57
收件人: dmlc/dgl <dgl@noreply.github.com>
抄送: Ziqiao Meng <ziqiao.meng@mail.utoronto.ca>; Author <author@noreply.github.com>
主题: Re: [dmlc/dgl] [Example] metapath2vec (#764)


Hi, I read some code in this repo, but it seems the sampler.py is not used. Now, this may be a Deepwalk implementation instead of metapath2vec?

―

**Comment by Theheavens** [comment]
In the examples/pytorch/metapath2vec/reading_data.py, 
And in the 117th line, I think you want to ignore the situation when the word is the centered word.
But maybe the variable j is not the centered word.

**Pull request merged**